### ISSUE TITLE
project saved in gpkg does not work

### ISSUE THREAD

**Issue by pigreco** [qgis/QGIS#37440]
A project file saved in a gpkg will not work if the file is renamed and moved to another folder.

how to reproduce it:

1. create a gpkg file from the browser panel
2. add a vector taken from other gpkg
3. thematize and save project in the gpkg;
4. close QGIS
5. move and rename the gpkg file;
6. start QGIS, create connection in the Browser panel
7. double click in the project in the gpkg
8. Error message is displayed, the carrier cannot be found

![image](https://user-images.githubusercontent.com/7631137/85919239-e7e76580-b869-11ea-8058-ec6f34177d09.png)

the layer has an incorrect link

[test.zip](https://github.com/qgis/QGIS/files/4840246/test.zip)

## OSGeo4W64 win10 - QGIS 3.14 Pi

![image](https://user-images.githubusercontent.com/7631137/85919296-47de0c00-b86a-11ea-82a4-6b8847204788.png)

**Comment by elpaso** [comment]
I would say this is a won't fix.

**Comment by agiudiceandrea** [comment]
Hi @pigreco, I'd like to test this issue, but I didn't quite understand how to proceed.
Should the attached file working or not working? It already contains a project and a vector layer.
Which gpkg should be moved and renamed: the one created by the browser panel or the "other gpkg"?

**Comment by pigreco** [comment]
> Ciao @pigreco , vorrei provare questo problema, ma non ho capito bene come procedere.
> Il file allegato dovrebbe funzionare o non funzionare? Contiene già un progetto e un livello vettoriale.
> Quale gpkg dovrebbe essere spostato e rinominato: quello creato dal pannello del browser o "l'altro gpkg"?

Hello,
the attached gpkg works and contains vector and project, but if you rename it and move it to another folder the project no longer works.

You have to move the renamed one.

But if you want to reproduce everything from scratch follow the bulleted list above.

**Comment by agiudiceandrea** [comment]
OK. For me, just renaming the gpkg file trigger the issue. If the gpkg file is not renamed, but only moved in another folder, I don't experience issues.

It seems to me this is due to the way the layer source is stored in the project file: "./test.gpkg|layername=Reg01012020_g_WGS84". The original gpkg file name is hardcoded (as well as the original layer name), while the path is relative.

Some new approach should be introduced to make the gpkg file name relative too, but only if the project is stored in the some gpkg file in which the layer is stored...
Or, at least, the Handle Unavailable Layer procedure could try to find the layer in the "current" gpkg file if the project is stored in it...

**Comment by gioman** [comment]
> OK. For me, just renaming the gpkg file trigger the issue. If the gpkg file is not renamed, but only moved in another folder, I don't experience issues.
> 
> It seems to me this is due to the way the layer source is stored in the project file: "./test.gpkg|layername=Reg01012020_g_WGS84". The original gpkg file name is hardcoded (as well as the original layer name), while the path is relative.
> 
> Some new approach could be introduced to make the gpkg file name relative too, but only if the project is stored in the some gpkg file in which the layer is stored...
> Or, at least, the Handle Unavailable Layer procedure could try to find the layer in the "current" gpkg file if the project is stored in it..

@pigreco @agiudiceandrea so this is either a won't fix or should be changed to a feature request, I guess...

**Comment by agiudiceandrea** [comment]
@gioman however the "gpkg project storage feature" was introduced as a way to "[create a **totally self-contained** GeoPackage which embeds a QGIS Project AND all the data used by the project.](https://www.qgis.org/en/site/forusers/visualchangelog38/index.html#feature-qgis-projects-inside-geopackage-files)" 

It seems to me the behavior reported by @pigreco is in contrast with what is arguable by the description.

Then, if it's not considered to be an issue, I think a statement should be added in the docs warning the user to not rename the self contained GeoPackage to avoid losing of the connections with the layers that itself contains.

**Comment by gioman** [comment]
@agiudiceandrea right, so after all is either a bug or the docs must be amended.

**Comment by nyalldawson** [comment]
I think this could possibly be fixed through the QgsPathResolver class -- at write time we could replace layer paths which match the gpkg project file with a special string, and then expand these out again at test time to the actual new gpkg project path...

**Comment by elpaso** [comment]
@nyalldawson yeah, a path rewrite was also my first though but it doesn't seem straightforward: the project writing to the storage happens by reading the temporary filesystem based QGZ (passed as a QDevice)  and storing it into the DB, we are missing an API that gets a (custom) path resolver from the storage class and sets it in the R/W context.

Maybe a better/easier approach could be to extend the broken layer handler to also search "self" storage and fix the paths.

**Comment by nyalldawson** [comment]
> yeah, a path rewrite was also my first though but it doesn't seem straightforward: the project writing to the storage happens by reading the temporary filesystem based QGZ (passed as a QDevice) and storing it into the DB, we are missing an API that gets a (custom) path resolver from the storage class and sets it in the R/W context.

Ok, I confirm your findings that the api is currently missing!

> Maybe a better/easier approach could be to extend the broken layer handler to also search "self" storage and fix the paths.

Maybe easier, but definitely not better :laughing:

**Comment by pigreco** [comment]
@nyalldawson @elpaso 
Please tell me how much it would cost to create the API to permanently fix this problem?

**Comment by cxcandid** [comment]
@pigreco @nyalldawson @elpaso @agiudiceandrea @gioman 
That was something that bothered me a long time ago and unfortunately still hasn't been resolved.
(see https://gis.stackexchange.com/questions/357344/trying-to-repair-layer-paths-on-opening-of-a-qgis-project-from-geopackage, https://gis.stackexchange.com/questions/330512/automatically-exchange-qgis-layer-datasources-during-project-opening).

As @nyalldawson mentioned, I used QgsPathResolver to automatically fix the invalid path names.
Today I uploaded my Plugin `changeGpkgPath` to the QGIS Plugins Repository.
I don't know if I've missed something, but this tiny plugin has proven to be very useful in the past.

**Comment by skprasadu** [comment]
I inspected the attached QGIS project/package locally.
Findings:
1. The attached `test.zip` contains `test.gpkg`.
2. `test.gpkg` contains a stored QGIS project named `test` in the `qgis_projects` table.
3. The stored project payload is a QGZ/ZIP archive. It contains:
   - `{23ddcdb2-3c8d-452a-b07c-81d593824b74}.qgs`
   - `{23ddcdb2-3c8d-452a-b07c-81d593824b74}.qgd`
4. The embedded `.qgs` project contains this datasource:
 
```text
   ./test.gpkg|layername=Reg01012020_g_WGS84
```
5. From the original extracted location, that relative datasource resolves to test.gpkg, so the layer source is available.
6. In a rename simulation where the package is copied/renamed to renamed.gpkg, the embedded project still resolves the layer datasource as:

```test
./test.gpkg
```
That points to a sibling file named test.gpkg, not to the renamed package renamed.gpkg.

This does not prove the root cause, but it narrows the failure mode to GeoPackage project relocation: the project payload is stored inside the GeoPackage and moves with it, while the embedded layer datasource remains filename-bound to the original GeoPackage name.

A useful next check would be:

* Rename test.gpkg to renamed.gpkg and open the stored project with no test.gpkg beside it.
* Then place a copy named test.gpkg beside renamed.gpkg and retry. If the layer becomes available, that confirms the stale filename reference.
